diff --git a/edgediscovery/dial.go b/edgediscovery/dial.go
index 1bbf59c3..aebbf312 100644
--- a/edgediscovery/dial.go
+++ b/edgediscovery/dial.go
@@ -1,14 +1,110 @@
 package edgediscovery
 
 import (
+	"bufio"
 	"context"
 	"crypto/tls"
+	"fmt"
+	"io"
 	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
 	"time"
 
 	"github.com/pkg/errors"
+	"golang.org/x/net/proxy"
 )
 
+type httpsProxyDialer string
+
+func (d httpsProxyDialer) DialContext(ctx context.Context, network, address string) (conn net.Conn, err error) {
+	defer func() {
+		if conn != nil && err != nil {
+			_ = conn.Close()
+			conn = nil
+		}
+	}()
+	if network != "tcp" {
+		return conn, fmt.Errorf("unsupported network for proxy '%s', support tcp", network)
+	}
+	httpsProxyUrl, err := url.Parse(string(d))
+	if err != nil {
+		return conn, fmt.Errorf("invalid proxy URL '%s', %w", string(d), err)
+	}
+	if httpsProxyUrl.Scheme != "http" {
+		return conn, fmt.Errorf("unsupported protocol for proxy '%s', support http://", string(d))
+	}
+	proxyHost := httpsProxyUrl.Host
+	if httpsProxyUrl.Port() == "" {
+		proxyHost = net.JoinHostPort(proxyHost, "80")
+	}
+	dialer := net.Dialer{}
+	conn, err = dialer.DialContext(ctx, network, proxyHost)
+	if err != nil {
+		return conn, fmt.Errorf("failed to dial proxy: %w", err)
+	}
+	host, portStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return conn, fmt.Errorf("invalid address to proxy: %s", address)
+	}
+	port, err := net.LookupPort("tcp", portStr)
+	if err != nil {
+		return conn, fmt.Errorf("invalid address to proxy: %s", address)
+	}
+	host = net.JoinHostPort(host, strconv.Itoa(port))
+	request := http.Request{
+		Method: "CONNECT",
+		URL: &url.URL{
+			Host: host,
+		},
+		Proto:      "HTTP/1.1",
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		Header: map[string][]string{
+			"User-Agent": {"cloudflared"},
+		},
+		Host: host,
+	}
+	connectCtx, cancel := context.WithTimeout(ctx, 1*time.Minute)
+	defer cancel()
+	var connectErr error
+	connectDone := make(chan struct{})
+	var response *http.Response
+	go func() {
+		defer close(connectDone)
+		err = request.Write(conn)
+		if err != nil {
+			connectErr = fmt.Errorf("failed to send CONNECT request to proxy: %w", err)
+			return
+		}
+		response, err = http.ReadResponse(bufio.NewReader(&io.LimitedReader{R: conn, N: 4 * 1024 * 1024}), &request)
+		if err != nil {
+			connectErr = fmt.Errorf("failed to receive http CONNECT response from proxy: %w", err)
+			return
+		}
+	}()
+	select {
+	case <-connectCtx.Done():
+		_ = conn.Close()
+		<-connectDone
+		return conn, fmt.Errorf("failed to CONNECT to proxy: %w", connectCtx.Err())
+	case <-connectDone:
+	}
+	if connectErr != nil {
+		return conn, connectErr
+	}
+	if response.StatusCode != 200 {
+		return conn, fmt.Errorf("proxy return %d response for CONNECT request", response.StatusCode)
+	}
+	return conn, nil
+}
+
+func (d httpsProxyDialer) Dial(network, address string) (net.Conn, error) {
+	return d.DialContext(context.Background(), network, address)
+}
+
 // DialEdge makes a TLS connection to a Cloudflare edge node
 func DialEdge(
 	ctx context.Context,
@@ -25,7 +121,31 @@ func DialEdge(
 	if localIP != nil {
 		dialer.LocalAddr = &net.TCPAddr{IP: localIP, Port: 0}
 	}
-	edgeConn, err := dialer.DialContext(dialCtx, "tcp", edgeTCPAddr.String())
+
+	httpsProxy := os.Getenv("https_proxy")
+	if httpsProxy == "" {
+		httpsProxy = os.Getenv("HTTPS_PROXY")
+	}
+	noProxy := os.Getenv("no_proxy")
+	if noProxy == "" {
+		noProxy = os.Getenv("NO_PROXY")
+	}
+
+	var proxiedDialer proxy.ContextDialer
+
+	if httpsProxy == "" {
+		proxiedDialer = &dialer
+	} else {
+		if noProxy == "" {
+			proxiedDialer = httpsProxyDialer(httpsProxy)
+		} else {
+			perHost := proxy.NewPerHost(httpsProxyDialer(httpsProxy), &dialer)
+			perHost.AddFromString(noProxy)
+			proxiedDialer = perHost
+		}
+	}
+
+	edgeConn, err := proxiedDialer.DialContext(dialCtx, "tcp", edgeTCPAddr.String())
 	if err != nil {
 		return nil, newDialError(err, "DialContext error")
 	}
